optuna/samplers/cmaes/sampler.py

```diff
@@ -0,0 +1,354 @@
+        if trial.number < self._n_startup_trials:
+            return {}
+
+        for name, distribution in intersection_search_space(study).items():
+            if distribution.single():
+                # `cma` cannot handle distributions that contain just a single value, so we skip
+                # them. Note that the parameter values for such distributions are sampled in
+                # `Trial`.
+                continue
+
+            if not isinstance(
+                    distribution,
+                    (
+                        optuna.distributions.UniformDistribution,
+                        optuna.distributions.LogUniformDistribution,
+                        optuna.distributions.DiscreteUniformDistribution,
+                        optuna.distributions.IntUniformDistribution,
+                    ),
+            ):
+                # Categorical distribution is unsupported.
+                continue
+            search_space[name] = distribution
+
+        if len(search_space) < 2:
+            self._logger.info("`CmaEsSampler` does not support optimization of 1-D search space. "
+                              "`{}` is used instead of `CmaEsSampler`.".format(
+                                  self._independent_sampler.__class__.__name__))
+            self._warn_independent_sampling = False
+            return {}
+
+        return search_space
+
+    def sample_relative(
+            self,
+            study: 'optuna.Study',
+            trial: 'optuna.structs.FrozenTrial',
+            search_space: Dict[str, BaseDistribution],
+    ) -> Dict[str, Any]:
+
+        if len(search_space) == 0:
+            return {}
+
+        completed_trials = [
+            t
+            for t in study.get_trials(deepcopy=False)
+            if t.state == TrialState.COMPLETE
+        ]
+        if len(completed_trials) < self._n_startup_trials:
+            return {}
+
+        ordered_keys = [key for key in search_space]
+        ordered_keys.sort()
+
+        optimizer = self._restore_or_init_optimizer(
+            completed_trials, search_space, ordered_keys
+        )
+
+        solution_trials = [
+            t
+            for t in completed_trials
+            if optimizer.generation == t.system_attrs.get("cma:generation", -1)
+        ]
+        if len(solution_trials) >= optimizer.population_size:
+            solutions = []  # type: List[Tuple[Any, float]]
+            for t in solution_trials[: optimizer.population_size]:
+                assert t.value is not None, "completed trials must have a value"
+                x = np.array([t.params[k] for k in ordered_keys])
+                solutions.append((x, t.value))
+
+            optimizer.tell(solutions)
+
+            optimizer_str = pickle.dumps(optimizer).hex()
+            study._storage.set_trial_system_attr(
+                trial._trial_id, "cma:optimizer", optimizer_str
+            )
+
+        # Caution: optimizer should update its seed value
```

Comment:
I assume that the previous CMA object is stored at trial 10.

* trial 10 : A multivariate gaussian distribution was updated. Then `cmaes.CMA` is serialized and stored in system_attr.
* trial 11 : Retrieving a cmaes.CMA object from trial 10, then sampling a parameter set.
* trial 12 : Retrieving a cmaes.CMA object from trial 10, then sampling a parameter set.
* trial 13 : Retrieving a cmaes.CMA object from trial 10, then sampling a parameter set.

If the same seed number is used, trials from 11 to 13 will sample the same parameter set. To avoid this problem, we have two approaches, set different seed numbers or set None. For reproducibility, I choose a former approach.